### PR TITLE
RFC: Stopping the yield typing madness with StreamingExecutionResult

### PR DESCRIPTION
**Pull request by schrockn**
## Summary & Motivation

One of the most frustrating things IMO programming in Dagster is the overreliance on `yield` for common use cases, and then the corresponding challenges in correctly typing that.

This proposes a new class `StreamingExecutionResult`, that seeks to remedy that.

The idea is that instead directly yielding from the `@asset` (or similar) decorator you instead return a `StreamingExecutionResult` that contains.

Unfortunately in python the ergonomics of lambda functions are inferior to a language like javascript, so this does require the creation of another function with a made up name.

However in exchange for that I think you get much better devex overall.

Before:

```python

    @multi_asset(outs={"one": AssetOut(), "two": AssetOut()})
    def multi():
        yield MaterializeResult(
            asset_key="one",
            metadata={"one": 1},
        )
        yield MaterializeResult(
            asset_key="two",
            metadata={"two": 2},
        )
```

you can instead:

```python


    @multi_asset(outs={"one": AssetOut(), "two": AssetOut()})
    def multi() -> StreamingExecutionResult:
        def _iter():
            yield MaterializeResult(
                asset_key="one",
                metadata={"one": 1},
            )
            yield MaterializeResult(
                asset_key="two",
                metadata={"two": 2},
            )
        return StreamingExecutionResult(_iter())
```

While adding `_iter` is slightly annoying, overall I think this is much more explict.

Name is up for debate since this can also consume lists of results, not just generators, so it is therefore not necessarily streaming. `StreamableExecutionResult` is an alternative name.

This leaves us in a world where we can type 100% of dagster assets and asset checks with either `MaterializeResult`, `AssetCheckResult` or `StreamableExecutionResult`. We can also add an `ExecutionResult` alias for the union of `MaterializeResult`, `AssetCheckResult` then add the inevitable `ObservationResult` or anything else in the future.

I think this is way more clear than direct yielding.

### CODE REVIEW

**Comment by schrockn** [comment]
Current dependencies on/for this PR:
* master
  * **PR #16906** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16906" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/16906?utm_source=stack-comment).

**Comment by schrockn** [comment]
`ExecutionResultStream` is another possibilty

**Comment by schrockn** [comment]
If we end up 100% standardizing on this "result" pattern in a later major version of Dagster, we can delete enormous amounts of type inference complexity.

**Comment by smackesey** [comment]
Certainly agree that having op compute funcs be either regular functions or generator functions introduces a lot of complexity internally that would be nice to expunge. Probably a source of bugs and general dev friction.

I'm less sold that this is an improvement in DX. While it is true that annotating generator functions can be unintuitive, if the user is trying to be type-safe they are probably going to want to annotate the wrapped generator function, so it's sort of kicking the can. That problem doesn't exist for generator expressions, but then we're just talking about the difference between this and this:

```
yield from (... gen exp ...)
# vs
return StreamingExecutionResult((... gen_exp ...))
``` 

>This leaves us in a world where we can type 100% of dagster assets and asset checks with either MaterializeResult, AssetCheckResult or StreamableExecutionResult. We can also add an ExecutionResult alias for the union of MaterializeResult, AssetCheckResult then add the inevitable ObservationResult or anything else in the future.

Suppose we're in a world where all the result objects are defined and we no longer return `Output` from assets. Then in `StreamableExecutionResult` world we would explain assets like:

>An asset compute function returns one or more result objects.

("one or more" because we still need to cover direct return of multiple results right?)

That's nice and simple. However, the user will discover that `StreamableExecutionResult` is different in kind from the other result objects, which correspond straightforwardly to operation results (`MaterializeResult`, `ObserveResult`, `AssetCheckResult`...).

Whereas if we keep direct `yield` the explanation is:

>An asset compute function returns or yields one or more result objects. 

And the result object ontology is considerably simpler.

Curious for others thoughts, but right now I'm on the side of keeping direct `yield`. I suspect some judicious refactors  could mitigate a lot of the internal complexity we currently have to deal with this.

**Comment by schrockn** [comment]
> I'm less sold that this is an improvement in DX. While it is true that annotating generator functions can be unintuitive, if the user is trying to be type-safe they are probably going to want to annotate the wrapped generator function, so it's sort of kicking the can.

To be clear I don't think I have _ever_ seen code in the wild that correctly type annotates the presence of yield. This at least compartmentalizes the lack of typing.

**Comment by schrockn** [comment]
> Suppose we're in a world where all the result objects are defined and we no longer return Output from assets. Then in StreamableExecutionResult world we would explain assets like:
>
> > An asset compute function returns one or more result objects.
> 
> ("one or more" because we still need to cover direct return of multiple results right?)

I think this is the most problematic part of the current name.

**Comment by schrockn** [comment]
> Curious for others thoughts, but right now I'm on the side of keeping direct yield. I suspect some judicious refactors could mitigate a lot of the internal complexity we currently have to deal with this.

To be clear, internal complexity wins are nice, but are _zero_ percent of the motivation here. This is all about consistency and understandable of the code that user's write.

**Comment by yuhan** [comment]
I'm supportive of this direction. I believe it makes it clear to both the internal system (so we could type it right) and users (so they know clearly when to yield vs return).

Particularly, I like that the proposed changes can result in return type consistency, meaning it could get us to answer "when yield vs return" by basically saying "always return". Concretely:
- In the beginner/essential paths, we don't need to talk about yielding at all - if you only want to return more than one object and don't care about the streaming behavior, just `return MaterializeResult(), MaterializeResult()`.
- Then, for users who care about the actual benefit of yields, `return StreamingExecutionResult` would make sense.
    ```
    yield from (... gen exp ...)
    # vs
    return StreamingExecutionResult((... gen_exp ...))
    ```
    Even it's just between these two, i find the latter much easier to explain.
        * in the current world, with the mix of `yield from`, `yield`, and `return`, I as a basic Dagster user could feel intimidated by the prospect of having to learn the difference.
        * whereas in the proposed latter case, I somehow feel like the consistency of returning `SomeResult` or `StreamingExecutionResult` makes the boundary of essential and advance more clear, as in, I as a basic user could feel comfortable not knowing anything about `StreamingExecutionResult`.

Re: yield typing madness, we hit the same issue in https://github.com/dagster-io/dagster/pull/16913#discussion_r1341640117. I'm curious if/how that can be fixed with this RFC.

**Comment by schrockn** [comment]
I also did not emphasize this, but I think a very important use case this makes easier is threading results from a library back to top-level. For example we have to tell people to use `yield from` for even the most basic usage of `@dbt_assets`. With the we could instead vanilla return this object and still get the streaming behavior. cc: @rexledesma

**Comment by jamiedemaria** [comment]
> This leaves us in a world where we can type 100% of dagster assets and asset checks with either MaterializeResult, AssetCheckResult or StreamableExecutionResult

does this mean that any value returned from an asset would go through `MaterializeResult`?

```python
@asset(
   io_manager_key="snowflake"
)
def my_snowflake_table():
    df = pd.DataFrame(...)
   return MaterializeResult(value=df) # instead of return df
```

**Comment by schrockn** [comment]
> does this mean that any value returned from an asset would go through `MaterializeResult`?

Yeah that is the maximalist vision of this.

**Comment by rexledesma** [comment]
In integration space, say in the `dagster-dbt` integration, are expecting:

1. The return type of `.stream() -> StreamingExecutionResult`? Or,
2. Do users have to wrap up the stream in this object?

I would prefer (1). But in order to not break users who are already using `yield from`, we should make `StreamingExecutionResult` implement `Iterator`, so that `yield from StreamExecutionResult(...)` continues to be treated as a container type, and iterates over its argument.

In other words, these should both work if we want to introduce this:


```python
yield from dbt.cli(...).stream()

return dbt.cli(...).stream()
```

**Comment by schrockn** [comment]
@rexledesma We would definitely introduce a different method, like `get_streaming_result()` or something

**Comment by sryza** [comment]
Before commenting on the specifics proposal, I'd like to raise a broader topic, which: what's our general attitude towards when it's appropriate to change these kinds of APIs?

For core APIs, I think that at a certain point, the question should become not "is this the best it can be?" but "is it good enough?" This is because there's a baseline cost to users any time we change how you do something in Dagster. At this point, there are mountains of code out there that people wrote using what we told them were the best practices for developing with Dagster. When we change those best practices, that code becomes partially obsolete, and the people responsible for or ramping up on those codebases end up caught between the old and new way of doing things.

So e.g. if we make a change to our APIs that make them 10% better, the net overall value to our users still might be negative.

To be clear, this isn't about when it's ok to make _breaking_ changes, which are obviously much more harmful. It's about changes to best practices.

---

Regarding the change itself, my _initial_ reaction is that it doesn't feel like the stated proposal makes a significant dent in the stated problem.

I.e. the fact that `def _iter():` doesn't get typed, and that the example still uses `yield`, undercuts the value of "most frustrating things IMO programming in Dagster is the overreliance on yield for common use cases, and then the corresponding challenges in correctly typing that."

I haven't had the chance to think deeply about this, so definitely don't take this as determined opposition to this change. So far am just evaluating the stated problem and solution at surface level. It seems very possible to me that there's a different way of stating the problem that would make it more clear that the proposed solution is an improvement.

**Comment by jamiedemaria** [comment]
> Yeah that is the maximalist vision of this 

ok. I can be convinced that that is a good place to go. I'm just a bit hesitant that the gains in non-io manager workflows that we made by introducing `MaterializeResult` will be undermined by allowing `value` to be anything. I'd want to see some different examples, especially the inter-op case where a single code location has some assets using an I/O manager and some not using an I/O manager

**Comment by PedramNavid** [comment]
I will second @sryza's comments here. It's not clear to me that there's significant improvement to the developer experience here. 

I would rather lean on documentation (both prose and API docs) to help soften this sharp edge rather than introduce a new concept, especially given the impact to existing implementations, even if it's not a breaking change. 

I sense that this isn't a big enough pain overall to justify the change, though I'm not against it, I just haven't personally seen evidence that the juice is worth the squeeze.

**Comment by schrockn** [comment]
Gotcha. I'll take another run at this next week since we are all neck-deep in launch week activities and work. I don't think I did an adequate job in framing this up, and I understand the hesitance given the framing.

As a preview I think this or something like it is pretty essential to complete the story that has been started with the introduction of `MaterializeResult` and `AssetCheckResult`, which, in my mind, are actually quite a bit more disruptive than the above proposed change.

**Comment by schrockn** [comment]
Started discussion based on this in 7247 in internal
